proposals/p2188.md

```diff
@@ -515,3 +655,11 @@ consistent with the behavior of `==` in the case where a struct is implicitly
 convertible to the class type. However, without a design for user-defined
 pattern matching and matching on dynamic type, there is a significant risk that
 this would conflict with the rules there, so it is deferred for now.
+
+### Type deduction
+
+This proposal does not cover type deduction, instead considering it to be a
+separate topic from pattern matching syntax, even though the semantic behavior
+of the two may be quite similar or identical.
```

Comment:
I don't think they're *necessarily* coupled. I think we are confident that we're going to do deduction when performing function calls to generic functions and when selecting an `impl`. I think we are confident that we're going to do pattern matching in a `match` statement and in a `let` or `var` binding. I don't think it's guaranteed that we'll do pattern matching in a function call -- it may be reasonable to restrict a function to have only a list of bindings as its parameters, which only needs pattern matching in a degenerate sense. So I think there might actually be no situations where we perform both type deduction and non-degenerate pattern matching.

I also think it's possible that we'll model the matching process for the two cases quite differently: for pattern matching, we have a top-down process where the pattern syntax determines what happens at each level (such as evaluating an expression pattern and performing an `==` comparison, or converting the scrutinee), whereas for type deduction I think it's likely we'll instead perform a symbolic evaluation of the type expression (the "pattern") and perform a value comparison between (symbolic) type values to find a set of substitutions. The difference is fairly subtle, but I think deduction ends up having stricter matching semantics than a pattern match -- identical after substitution versus a looser notion of "match".

Reply:
> So I think there might actually be no situations where we perform both type deduction and non-degenerate pattern matching.

It seems like variadic functions will be such a situation. Maybe we can say that semantically, variadics relate only to deduction, and not pattern matching (although I'm not sure of that), but syntactically they're going to appear in the parentheses as well as the square brackets.